src/ripple/app/misc/NegativeUNLVote.h

```diff
@@ -0,0 +1,189 @@
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_MISC_NEGATIVEUNLVOTE_H_INCLUDED
+#define RIPPLE_APP_MISC_NEGATIVEUNLVOTE_H_INCLUDED
+
+#include <ripple/beast/utility/Journal.h>
+
+namespace ripple {
+
+class Ledger;
+template <class Adaptor>
+class Validations;
+class RCLValidationsAdaptor;
+using RCLValidations = Validations<RCLValidationsAdaptor>;
+class SHAMap;
+namespace test {
+class NegativeUNLVoteInternal_test;
+class NegativeUNLVoteScoreTable_test;
+}  // namespace test
+
+/**
+ * Manager to create NegativeUNL votes.
+ */
+class NegativeUNLVote final
+{
+public:
+    /**
+     * A validator is considered unreliable if its validations is less than
+     * nUnlLowWaterMark in the last flag ledger period.
+     * An unreliable validator is a candidate to be disabled by the NegativeUNL
+     * protocol.
+     */
+    static constexpr size_t nUnlLowWaterMark = 128;  // 256 * 0.5;
+    /**
+     * An unreliable validator must have more than nUnlHighWaterMark validations
+     * in the last flag ledger period to be re-enabled.
+     */
+    static constexpr size_t nUnlHighWaterMark = 204;  //~256 * 0.8;
+    /**
+     * The minimum number of validations of the local node for it to
+     * participate in the voting.
+     */
+    static constexpr size_t nUnlMinLocalValsToVote = 230;  //~256 * 0.9;
+    /**
+     * We don't want to disable new validators immediately after adding them.
+     * So we skip voting for disabling them for 2 flag ledgers.
+     */
+    static constexpr size_t newValidatorDisableSkip = 512;  // 256 * 2;
+    /**
+     * We only want to put 25% of the UNL on the NegativeUNL.
+     */
+    static constexpr float nUnlMaxListed = 0.25;
+
+    /**
+     * Constructor
+     *
+     * @param myId the NodeID of the local node
+     * @param j log
+     */
+    NegativeUNLVote(NodeID const& myId, beast::Journal j);
+    ~NegativeUNLVote() = default;
+
+    using LedgerConstPtr = std::shared_ptr<Ledger const> const;
+
+    /**
+     * Cast our local vote on the NegativeUNL candidates.
+     *
+     * @param prevLedger the parent ledger
+     * @param unlKeys the trusted master keys
+     * @param validations the validation message container
+     * @param initialSet the set of transactions
+     */
+    void
+    doVoting(
+        LedgerConstPtr& prevLedger,
+        hash_set<PublicKey> const& unlKeys,
+        RCLValidations& validations,
+        std::shared_ptr<SHAMap> const& initialSet);
+
+    /**
+     * Notify NegativeUNLVote that new validators are added.
+     * So that they don't get voted to the NegativeUNL immediately.
+     *
+     * @param seq the current LedgerIndex when adding the new validators
+     * @param nowTrusted the new validators
+     */
+    void
+    newValidators(LedgerIndex seq, hash_set<NodeID> const& nowTrusted);
+
+private:
+    NodeID const myId_;
+    beast::Journal j_;
+    std::mutex mutex_;
```

Comment:
Member vars of type `mutex` should almost always be `mutable`

Reply:
Thanks, changed.